### PR TITLE
Create user token with Strategic IdAM

### PR DESCRIPTION
**Pull request by KarolJastrzebski**
Functional tests are ran against fully functional instance of the application and try to simulate a real user.

SAAT is run with SIDAM, whereas AAT uses TIDAM.

This pull request features:
- switchable implementations of obtaining JWT for both TIDAM and SIDAM
- new IdAM client-service in docker
- functional tests can be run against docker-composed environment

```bash
./gradlew clean assemble && docker-compose up --build
./gradlew functional
```

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/aat/java/uk/gov/hmcts/reform/coh/idam/AbstractIdamAuthentication.java [hunk 1]
```diff
@@ -45,19 +33,11 @@ public void init() {
...
-        token = client.lease(userId, role);
-        if (token != null && !"".equals(token)) {
-            log.info("Token refreshed");
-        } else {
-            log.warn("Empty token!");
-        }
+        token = newToken();
     }
 
+    abstract protected String newToken();
```
Comment (johnhosolirius) [inline src/aat/java/uk/gov/hmcts/reform/coh/idam/AbstractIdamAuthentication.java]: getNewToken

File: src/aat/java/uk/gov/hmcts/reform/coh/idam/IdamClient.java [hunk 2]
```diff
@@ -2,9 +2,13 @@
 
 public interface IdamClient {
 
-    void createAccount(String email, String role);
+    void createAccount(String email, String role, String password);
```
Comment (johnhosolirius) [inline src/aat/java/uk/gov/hmcts/reform/coh/idam/IdamClient.java]: Do like this -> void createAccount(String email, String password, String role);
More logical to have the password next to the username

File: src/aat/java/uk/gov/hmcts/reform/coh/idam/TacticalIdamAuthentication.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    @Override
+    protected String newToken() {
+        Integer userId = client.findUserByEmail(email);
+        if (userId == 0) {
+            client.createAccount(email, role, UUID.randomUUID().toString());
+            userId = client.findUserByEmail(email);
+        }
+        return client.lease(userId, role);
+    }
+}
```
Comment (johnhosolirius) [inline src/aat/java/uk/gov/hmcts/reform/coh/idam/TacticalIdamAuthentication.java]: Not sure why we need this implementation since Tidam also supports the "new" way

File: src/aat/resources/application.yaml [hunk 4]
```diff
@@ -29,7 +29,7 @@ spring:
 
 base-urls:
   idam-url: ${IDAM_API_URL:http://localhost:4501}
-  idam-user-email: ${IDAM_USER_EMAIL:test-citizen@example.com}
+  idam-user-email: ${IDAM_USER_EMAIL:test-citizen-1@example.com}
```
Comment (johnhosolirius) [inline src/aat/resources/application.yaml]: Why the change in user name?

File: src/aat/java/uk/gov/hmcts/reform/coh/idam/AbstractIdamAuthentication.java (hunk 1 above)
Comment (KarolJastrzebski) [reply src/aat/java/uk/gov/hmcts/reform/coh/idam/AbstractIdamAuthentication.java]: Done

File: src/aat/java/uk/gov/hmcts/reform/coh/idam/IdamClient.java (hunk 2 above)
Comment (KarolJastrzebski) [reply src/aat/java/uk/gov/hmcts/reform/coh/idam/IdamClient.java]: Done.
